x-pack/plugin/profiling/src/main/java/org/elasticsearch/xpack/profiling/TransportGetStackTracesAction.java

```diff
@@ -526,14 +526,25 @@ public void onStackTraceResponse(MultiGetResponse multiGetItemResponses) {
                         if (stackTracePerId.putIfAbsent(id, stacktrace) == null) {
                             totalFrames.addAndGet(stacktrace.frameIds.size());
                             stackFrameIds.addAll(stacktrace.frameIds);
-                            executableIds.addAll(stacktrace.fileIds);
+                            forNativeAndKernelFrames(stacktrace, executableId -> executableIds.add(executableId));
                         }
                     }
                 }
             }
             mayFinish();
         }
 
+        private static void forNativeAndKernelFrames(StackTrace stacktrace, java.util.function.Consumer<String> consumer) {
```

Comment:
As this method operates entirely on `StackTrace` how about turning it into a method on that class?

```java
    public List<String> nativeAndKernelFileIds() {
        // just kept here to keep the code self-contained - move to class-level constants
        final int NATIVE_FRAME_TYPE = 3;
        final int KERNEL_FRAME_TYPE = 4;
        List<String> collectedFileIds = new ArrayList<>();
        for (int i = 0; i < fileIds.size(); i++) {
            Integer frameType = typeIds.get(i);
            if (frameType != null && (frameType == NATIVE_FRAME_TYPE || frameType == KERNEL_FRAME_TYPE)) {
                collectedFileIds.add(fileIds.get(i));
            }
        }
        return collectedFileIds;
    }
```

Reply:
Sounds good, done.